piet/src/gradient.rs

```diff
@@ -50,3 +54,158 @@ pub struct GradientStop {
+        GradientStops::to_vec(stops)
+    }
+}
+
+impl<'a> GradientStops for (Color, Color, Color) {
+    fn to_vec(self) -> Vec<GradientStop> {
+        let stops: &[Color] = &[self.0, self.1, self.2];
+        GradientStops::to_vec(stops)
+    }
+}
+
+impl<'a> GradientStops for (Color, Color, Color, Color) {
+    fn to_vec(self) -> Vec<GradientStop> {
+        let stops: &[Color] = &[self.0, self.1, self.2, self.3];
+        GradientStops::to_vec(stops)
+    }
+}
+
+impl<'a> GradientStops for (Color, Color, Color, Color, Color) {
+    fn to_vec(self) -> Vec<GradientStop> {
+        let stops: &[Color] = &[self.0, self.1, self.2, self.3, self.4];
+        GradientStops::to_vec(stops)
+    }
+}
+
+impl<'a> GradientStops for (Color, Color, Color, Color, Color, Color) {
+    fn to_vec(self) -> Vec<GradientStop> {
+        let stops: &[Color] = &[self.0, self.1, self.2, self.3, self.4, self.5];
+        GradientStops::to_vec(stops)
+    }
+}
+
+impl UnitPoint {
+    pub const TOP_LEFT: UnitPoint = UnitPoint::new(0.0, 0.0);
+    pub const TOP: UnitPoint = UnitPoint::new(0.5, 0.0);
+    pub const TOP_RIGHT: UnitPoint = UnitPoint::new(1.0, 0.0);
+    pub const LEFT: UnitPoint = UnitPoint::new(0.0, 0.5);
+    pub const CENTER: UnitPoint = UnitPoint::new(0.5, 0.5);
+    pub const RIGHT: UnitPoint = UnitPoint::new(1.0, 0.5);
+    pub const BOTTOM_LEFT: UnitPoint = UnitPoint::new(0.0, 1.0);
+    pub const BOTTOM: UnitPoint = UnitPoint::new(0.5, 1.0);
+    pub const BOTTOM_RIGHT: UnitPoint = UnitPoint::new(1.0, 1.0);
+
+    /// Create a new UnitPoint.
+    ///
+    /// The `u` and `v` coordinates describe the point, with 0.0 being
+    /// left and top, and 1.0 being right and bottom, respectively.
+    pub const fn new(u: f64, v: f64) -> UnitPoint {
+        UnitPoint { u, v }
+    }
+    /// Given a rectangle, resolve the point within the rectangle.
+    pub fn resolve(&self, rect: Rect) -> Point {
+        Point::new(
+            rect.x0 + self.u * (rect.x1 - rect.x0),
+            rect.y0 + self.v * (rect.y1 - rect.y0),
+        )
+    }
+}
+
+impl LinearGradient {
+    /// Create a new linear gradient.
+    ///
+    /// The `start` and `end` coordinates are [`UnitPoint`] coordinates, relative
+    /// to the geometry of the shape being drawn.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use piet::{Color, FillRule, RenderContext, LinearGradient, UnitPoint};
+    /// use piet::kurbo::{Circle, Point};
+    ///
+    /// # let mut render_ctx = piet::NullRenderContext::new();
+    /// let circle = Circle::new(Point::new(100.0, 100.0), 50.0);
+    /// let gradient = LinearGradient::new(
+    ///     UnitPoint::TOP,
+    ///     UnitPoint::BOTTOM,
+    ///     (Color::WHITE, Color::BLACK)
+    /// );
+    /// render_ctx.fill(circle, &gradient, FillRule::NonZero);
+    /// ```
+    ///
+    /// [`UnitPoint`]: struct.UnitPoint.html
+    pub fn new(start: UnitPoint, end: UnitPoint, stops: impl GradientStops) -> LinearGradient {
```

Comment:
that this function is essentially variadic makes me wonder if life might not be easier if we had a `linear_grad!` macro; took a quick look at this though and it isn't totally trivial, especially if we want some of the nice stuff like handling both `Color`s and `GradientStop`s.

Reply:
I think what's there is a pretty good compromise. I was trying to make some form of `&[color1, color2]` work, but that syntax actually gives you a `&[Color; 2]` type, not a slice. It's likely Rust will get better at this sort of thing, with integer generics or whatnot.

I'm against the use of macros for minor syntax sugaring; it makes the code a *lot* harder to follow. So my strong inclination is to play with types first, and only use macros if there's a lot of repetitive boilerplate, or if there's an invariant that's easier to enforce.